toolchains/python/python.bzl

```diff
@@ -200,3 +200,184 @@ def nixpkgs_python_configure(
+
+    repository_ctx.file("BUILD.bazel", content)
+
+    # 3. generate BUILD.bazel file,
+    # ... _and_ the symlinks
+    # 4. Generate dummy WORKSPACE
+    # repository_ctx.file("WORKSPACE", "")
+
+    repository_ctx.file("python_module.bzl", """
+def _python_module_impl(ctx):
+    import_depsets = []
+    store = ctx.file.store_path
+    runfiles = ctx.runfiles(files = [store])
+
+    for dep in ctx.attr.deps:
+        runfiles = runfiles.merge(dep[DefaultInfo].data_runfiles)
+        import_depsets.append(dep[PyInfo].imports)
+
+    # HACK(danny): for some unforunate reason, short_path returns ../ when operating in external
+    # repositories. I don't know why. It breaks rules_python's assumptions though.
+    fixed_path = store.short_path[3:]
+    import_path = "/".join([ctx.workspace_name, store.short_path])
+
+    return [
+        DefaultInfo(
+            files = depset(ctx.files.files),
+            default_runfiles = ctx.runfiles(ctx.files.files, collect_default = True),
+        ),
+        PyInfo(
+            imports = depset(direct = [import_path], transitive = import_depsets),
+            transitive_sources = depset(transitive = [
+                dep[PyInfo].transitive_sources
+                for dep in ctx.attr.deps
+            ]),
+        ),
+    ]
+
+python_module = rule(
+    implementation = _python_module_impl,
+    attrs = {
+        "store_path": attr.label(
+            allow_single_file = True,
+            doc = "nix store path of python package",
+        ),
+        "files": attr.label_list(
+            allow_files = True,
+        ),
+        "deps": attr.label_list(
+            providers = [PyInfo],
+        ),
+    },
+    executable = False,
+    test = False,
+)
+""",
+    )
+
+    # 5. generate //:requirements.bzl for later import.
+    repository_ctx.file("requirements.bzl", """
+def requirement(package_name):
+    return "@{}//:{{}}".format(package_name)
+""".format(repository_ctx.name),
+    )
+
+    # TODO: make it lazy in the packages themselves ?
+
+
+_nixpkgs_python_repository = repository_rule(
+    _nixpkgs_python_repository_impl,
+    attrs = {
+        "json_deps": attr.label(),
+    },
+)
+
+
+def nixpkgs_python_repository(
+        name,
+        repository = None,
+        repositories = {},
+        nix_file = None,
+        nix_file_deps = [],
+        quiet = False,
+        ):
+    """Define a collection of python modules.
+
+    Args:
+      name: The name for the created module set.
+      repository: See [`nixpkgs_package`](#nixpkgs_package-repository).
+      repositories: See [`nixpkgs_package`](#nixpkgs_package-repositories).
+      nix_file: See [`nixpkgs_package`](#nixpkgs_package-nix_file).
+      nix_file_deps: See [`nixpkgs_package`](#nixpkgs_package-nix_file_deps).
+    """
+
+    generated_deps_name = "_generated_{}_deps".format(name)
+
+    nixpkgs_package(
+        name = generated_deps_name,
+        nix_file_content = """
```

Comment:
Since the question came up in the Q/A. You can [see here](https://github.com/tweag/rules_nixpkgs/blob/3748711f010bd191b0867344d615f9dc98b6eaf9/toolchains/cc/cc.bzl#L429-L430) how this can be done with a `nix_file` attribute and a file stored in `rules_nixpkgs_python`.